### PR TITLE
[office-js-preview] (Shared runtime) Add before close notification preview APIs

### PR DESCRIPTION
**Pull request by alison-mk**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: Beta URL: https://appsforoffice.microsoft.com/lib/beta/hosted/office.debug.js
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/office-js-preview/index.d.ts [hunk 1]
```diff
@@ -573,6 +573,59 @@ declare namespace Office {
         onVisibilityModeChanged(
             listener: (message: VisibilityModeChangedMessage) => void,
         ): Promise<() => Promise<void>>;
+
+        /**
+         * Represents a notification dialog that can appear when the user closes a document, before allowing the document to close.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts:578]:
```suggestion
         * Represents a modal notification dialog that can appear when the user attempts to close a document. The document will  not close until the user responds..
```

File: types/office-js-preview/index.d.ts [hunk 2]
```diff
@@ -573,6 +573,59 @@ declare namespace Office {
...
+         * 
+         * @remarks
+         * [Api set: SharedRuntime BETA (PREVIEW ONLY)]
+         * @beta
+        */
+        beforeDocumentCloseNotification: BeforeDocumentCloseNotification;
+    }
+    /**
+     * Represents a notification dialog that can appear when the user closes a document, before allowing the document to close.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts:587]:
```suggestion
     * Represents a modal notification dialog that can appear when the user attempts to close a document. The document will not close until the user responds..
```

File: types/office-js-preview/index.d.ts [hunk 3]
```diff
@@ -573,6 +573,59 @@ declare namespace Office {
...
+     * Represents a notification dialog that can appear when the user closes a document, before allowing the document to close. 
+     * The notification dialog will allow the user to confirm the request to close the document or cancel the request to close the document.
+     * 
+     * @remarks
+     * [Api set: SharedRuntime BETA (PREVIEW ONLY)]
+     * @beta
+    */
+    interface BeforeDocumentCloseNotification {
+        /**
+         * Enable a notification dialog that appears when the user closes a document, before allowing the document to close.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts:596]:
```suggestion
         * Enable a modal notification dialog that appears when the user attempts to close a document. The document will  not close until the user responds..
```

File: types/office-js-preview/index.d.ts [hunk 4]
```diff
@@ -573,6 +573,59 @@ declare namespace Office {
...
+         * This notification dialog asks the user to confirm the request to close the document, or allows the user to cancel the request to close the document.
+         * 
+         * @remarks
+         * [Api set: SharedRuntime BETA (PREVIEW ONLY)]
+         * @beta
+         */
+        enable(): Promise<void>;
+
+        /**
+         * Disable a notification dialog that appears when the user closes a document, before allowing the document to close.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts:606]:
```suggestion
         * Prevents the notification dialog from appearing when the user attempts to close a document.
```

File: types/office-js-preview/index.d.ts [hunk 5]
```diff
@@ -573,6 +573,59 @@ declare namespace Office {
...
+         * [Api set: SharedRuntime BETA (PREVIEW ONLY)]
+         * @beta
+         */
+        disable(): Promise<void>;
+
+        /**
+         * Adds a listener that detects when the `BeforeDocumentCloseNotification` close operation is cancelled. 
+         * This listener will be triggered if one of the following conditions is met:
+         * 1. When the notification dialog is open, the end user clicks the **Don't close** button within the dialog, clicks the Close button in the upper right corner of the dialog, or presses the Esc key.
+         * 2. When the local add-in calls the `enable` method on `BeforeDocumentCloseNotification`.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts:618]:
```suggestion
         * 2. When the add-in calls the `enable` method on `BeforeDocumentCloseNotification` object.
```

File: types/office-js-preview/index.d.ts [hunk 6]
```diff
@@ -573,6 +573,59 @@ declare namespace Office {
...
+         * @param listener The listener function that is called when the before close notification dialog is cancelled.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts:619]:
```suggestion
         * @param listener The listener function that is called when the dialog is cancelled.
```

**Review by Rick-Kirkham — CHANGES_REQUESTED** [review]

File: types/office-js-preview/index.d.ts (hunk 1 above)
Comment (alison-mk) [reply types/office-js-preview/index.d.ts:578]: I kept getting errors from GitHub when trying to commit your suggestions directly, so I added them in a new commit.

Comment (Rick-Kirkham) [reply types/office-js-preview/index.d.ts:578]: Yeah. You can't accept suggestions on DefinitelyTyped for some reason. I only use them as an easy way of describing what I think the change should be.

File: types/office-js-preview/index.d.ts [hunk 7]
```diff
@@ -573,6 +573,59 @@ declare namespace Office {
...
+         * Prevents the notification dialog from appearing when the user attempts to close a document.
+         * 
+         * @remarks
+         * [Api set: SharedRuntime BETA (PREVIEW ONLY)]
+         * @beta
+         */
+        disable(): Promise<void>;
+
+        /**
+         * Adds a listener that detects when the `BeforeDocumentCloseNotification` close operation is cancelled.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts:615]: Please use "handler" or "event handler" in place of "listener" throughout this API, including the signature. 

(For some reason this comment didn't get saved when I reviewed last night.)

**Review by Rick-Kirkham — CHANGES_REQUESTED** [review]
One more thing that I noted last night, but for some reason it wasn't saved.

File: types/office-js-preview/index.d.ts (hunk 7 above)
Comment (alison-mk) [reply types/office-js-preview/index.d.ts:615]: This API follows the same pattern as onVisibilityModeChanged (directly preceding, see line 573), including "listener" in the signature. Do you still think that I should change listener to handler for onCloseActionCancelled?

Comment (Rick-Kirkham) [reply types/office-js-preview/index.d.ts:615]: > This API follows the same pattern as onVisibilityModeChanged (directly preceding, see line 573), including "listener" in the signature. Do you still think that I should change listener to handler for onCloseActionCancelled?

@alison-mk I see. That's not the correct term, and the conceptual topic for onVisbilityModeChanged ([Handle the visibility changed event](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/show-hide-add-in#handle-the-visibility-changed-event)) correctly calls it a "handler". Let's not introduce an inconsistency. So, you can either change both the events to "handler" in this PR or leave them both as "listener" now and create a work item for the team to fix this in the future. Let me know if you prefer the latter alternative and I'll approve. Thanks.

Comment (alison-mk) [reply types/office-js-preview/index.d.ts:615]: I feel uncertain about changing a type definition signature without consulting an engineer. I'm not sure what the ripple effects might be. I can change "listener" to "event handler" in the description though.

Comment (Rick-Kirkham) [reply types/office-js-preview/index.d.ts:615]: > I feel uncertain about changing a type definition signature without consulting an engineer. I'm not sure what the ripple effects might be. I can change "listener" to "event handler" in the description though.

OK. I'll approve that, and I'll make a work item to fix this.

**Review by Rick-Kirkham — APPROVED** [review]

**Pull request merged**